File: 프로그래머스/unrated/172927. 광물 캐기/광물 캐기.py
def solution(picks, minerals):
    answer = 0
    
    # 쓸 수 있는 곡괭이 수
    picks_cnt = 0
    for x in picks:
        picks_cnt += x
    
    # 캘 수 있는 광물 수가 더 많을 경우 
    if picks_cnt *5 < len(minerals):
        # 못 캐는 광물은 버린다
        minerals = minerals[:picks_cnt*5]
    # minerals 에서 광물 개수 정렬
    minerals_sorted = [[0,0,0] for _ in range(len(minerals)//5 + 1)]
    for i in range(len(minerals)):
        if minerals[i]=="diamond":
            minerals_sorted[i//5][0] +=1
        elif minerals[i]== "iron":
            minerals_sorted[i//5][1] +=1
        elif minerals[i]=="stone":
            minerals_sorted[i//5][2] +=1
    
    minerals_sorted = sorted(minerals_sorted, key = lambda x:(-x[0], -x[1], -x[2]))
    for x in minerals_sorted:
        d,i,s = x
        for j in range(len(picks)):
            if j == 0 and picks[j]>0:
                picks[j]-=1
                answer += d + i + s
                break
            elif j == 1 and picks[j]>0:
                picks[j]-=1
                answer += (5*d) + i + s
                break
            elif j == 2 and picks[j]>0:
                picks[j]-=1
                answer += (25*d) + (5*i) + s
                break
                
    return answer

File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.js
function solution(picks, minerals) {
    var answer = 0;
    let count = 0 // 5개 묶음 카운트
    let groups = 0 // 5개씩 묶음 값
    let fatigue = 0 // 피로도 
    let dPck = picks[0]
    let iPck= picks[1]
    let sPck = picks[2]
    const sortedMineralGroups = []
    
    // 곡괭이 부족해서 캘 수 없는 광물 없애기
    const len = dPck * 5 + iPck * 5 + sPck * 5
    minerals.splice(len)

    for(let i = 0; i < minerals.length; i++){
        let tmp = 0 
        if(minerals[i] === "diamond") tmp = 100;
        if(minerals[i] === "iron") tmp = 10;
        if(minerals[i] === "stone") tmp = 1;
        
        groups += tmp;
        count++;
        
        if(count === 5){
            sortedMineralGroups.push(groups)
            groups = 0
            count = 0
        }
    }
    
    if(count) sortedMineralGroups.push(groups)
    sortedMineralGroups.sort((a, b) => b - a)
    

    let idx = 0
    
   
    
    while((dPck !==0 || iPck!==0 || sPck!==0)  && idx < sortedMineralGroups.length){
        const currMineral = sortedMineralGroups[idx]
        const diaCnt = Math.floor(currMineral/100);
        const ironCnt = Math.floor((currMineral%100) / 10);
        const stoneCnt = currMineral % 10;
        idx++;
        if(dPck !== 0){
            fatigue += diaCnt + ironCnt + stoneCnt
            dPck--;
            continue;
        }
        
        if(iPck !== 0){
            fatigue += diaCnt * 5 + ironCnt + stoneCnt
            iPck--;
            continue
        }
        
        if(sPck !== 0){
            fatigue += diaCnt * 25 + ironCnt * 5 + stoneCnt
            sPck--;
        }
    }
    
    return fatigue;
}

